activation/nipost.go

```diff
@@ -310,50 +323,71 @@ func withConditionalTimeout(ctx context.Context, timeout time.Duration) (context
 	}
 	logger := nb.log.With(log.ZContext(ctx), zap.String("poet", client.Address()))
 
-	logger.Debug("querying for poet pow parameters")
-	powCtx, cancel := withConditionalTimeout(ctx, nb.poetCfg.RequestTimeout)
-	defer cancel()
-	powParams, err := client.PowParams(powCtx)
-	if err != nil {
-		return nil, &PoetSvcUnstableError{msg: "failed to get PoW params", source: err}
+	// FIXME: remove support for deprecated poet PoW
+	auth := PoetAuth{
+		PoetCert: certifier.GetCertificate(client.Address()),
 	}
-
-	logger.Debug("doing pow with params", zap.Any("pow_params", powParams))
-	startTime := time.Now()
-	nonce, err := shared.FindSubmitPowNonce(
-		ctx,
-		powParams.Challenge,
-		challenge,
-		nb.signer.NodeID().Bytes(),
-		powParams.Difficulty,
-	)
-	metrics.PoetPowDuration.Set(float64(time.Since(startTime).Nanoseconds()))
-	if err != nil {
-		return nil, fmt.Errorf("running poet PoW: %w", err)
+	if auth.PoetCert == nil {
+		logger.Info("missing poet cert - falling back to PoW")
+		powCtx, cancel := withConditionalTimeout(ctx, nb.poetCfg.RequestTimeout)
+		defer cancel()
+		powParams, err := client.PowParams(powCtx)
+		if err != nil {
+			return nil, &PoetSvcUnstableError{msg: "failed to get PoW params", source: err}
+		}
+		logger.Debug("doing pow with params", zap.Any("pow_params", powParams))
+		startTime := time.Now()
+		nonce, err := shared.FindSubmitPowNonce(
+			ctx,
+			powParams.Challenge,
+			challenge,
+			nb.signer.NodeID().Bytes(),
+			powParams.Difficulty,
+		)
+		metrics.PoetPowDuration.Set(float64(time.Since(startTime).Nanoseconds()))
+		if err != nil {
+			return nil, fmt.Errorf("running poet PoW: %w", err)
+		}
+		auth.PoetPoW = &PoetPoW{
+			Nonce:  nonce,
+			Params: *powParams,
+		}
+	} else {
+		logger.Info("registering with a certificate", zap.Binary("cert", auth.PoetCert))
 	}
 
 	logger.Debug("submitting challenge to poet proving service")
 
 	submitCtx, cancel := withConditionalTimeout(ctx, nb.poetCfg.RequestTimeout)
 	defer cancel()
-	round, err := client.Submit(submitCtx, deadline, prefix, challenge, signature, nb.signer.NodeID(), PoetPoW{
-		Nonce:  nonce,
-		Params: *powParams,
-	})
-	if err != nil {
-		return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
-	}
+	var round *types.PoetRound
 
+	for try := 0; try < 2; try++ {
+		round, err = client.Submit(submitCtx, deadline, prefix, challenge, signature, nb.signer.NodeID(), auth)
+		if err == nil {
+			break
+		}
+		if errors.Is(err, ErrUnathorized) && try == 0 {
+			logger.Warn("failed to submit challenge as unathorized - recertifying", zap.Error(err))
+			auth.PoetCert, err = certifier.Recertify(ctx, client)
+			if err != nil {
+				return nil, &PoetSvcUnstableError{msg: "failed to regenerate poet certificate", source: err}
+			}
+		} else {
+			return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
+		}
+	}
```

Comment:
This seems a bit too clever to me. Would it be possible to first check if we have a valid certificate, if not re-certify (return on an error) and then trying to submit (and return on an error)?

Reply:
Why is it "clever"? The code:
1. calls /submit with a certificate it already has in the DB
2. re-generates the certificate if the existing one turned out to be invalid

The code could be rewritten to avoid the loop :shrug:.

>Would it be possible to first check if we have a valid certificate, if not re-certify

I want to avoid an extra call to /info to obtain the certifier info before every /submit. In most cases, the certificate should be valid so the code optimistically assumes it and just re-certifies on an auth error.

Reply:
I added a new comment for this - basically the code is a bit difficult to understand and I think it would be easier to put this logic into `client.Submit` (or some equivalent method on the certifier), i.e. fetching the certificate from DB and rectifying if needed

Reply:
The code is updated as you proposed in the other comment.